client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java

```diff
@@ -361,9 +355,16 @@ private Date getDate(int columnIndex, DateTimeZone localTimeZone)
 
     private static Date parseDate(String value, DateTimeZone localTimeZone)
     {
-        long millis = DATE_FORMATTER.withZone(localTimeZone).parseMillis(String.valueOf(value));
-        if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
-            return new Date(millis);
+        try {
+            long millis = DATE_FORMATTER.withZone(localTimeZone).parseMillis(String.valueOf(value));
+            if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
+                return new Date(millis);
+            }
+        }
+        catch (IllegalInstantException ignored) {
+            // https://joda-time.sourceforge.net/faq.html#illegalinstant
+            LocalDate localDate = DATE_FORMATTER.parseLocalDate(String.valueOf(value));
+            return new Date(localDate.toDateTimeAtStartOfDay(localTimeZone).getMillis());
```

Comment:
- shouldn't this be done only if `millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND`?
 the code block below is supposed to be handling `!(millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND)` cases, but it's no longer the case

- why do we parse with two different ways? can't we use `DATE_FORMATTER.parseLocalDate` always?
  if there is a good reason for this, there should be a code comment explaining it